<launch>

    <node name="brs_debug_status" pkg="karcher" type="brs_debug_status" output="screen"/>
    <node name="brs_motor_current" pkg="karcher" type="brs_motor_current"/>
    <node name="brs_dt" pkg="karcher" type="brs_dt" output="screen"/>
    <node name="brs_checking" pkg="karcher" type="brs_checking"/>
    <node name="brs_node_check" pkg="karcher" type="brs_node_check"/>
<!--    <node name="brs_odom" pkg="karcher" type="brs_odom"/>
-->
<!--    <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
        <param name="output_frame" value="odom"/>
        <param name="freq" value="20.0"/>
        <param name="sensor_timeout" value="3.0"/>
        <param name="odom_used" value="true"/>
        <param name="imu_used" value="true"/>
        <param name="vo_used" value="false"/>
        <param name="debug" value="false"/>
        <param name="self_diagnose" value="false"/>
    </node>

    <node pkg="tf" type="static_transform_publisher" name="imu_tf_broadcaster" args="0.2 0 0.5 0 0 0 base_link imu_link 50" />

    <node pkg="tf" type="static_transform_publisher" name="base_tf_broadcaster" args="0 0 0 0 0 0 base_footprint base_link 50" />
-->
    <node name="brs_sense" pkg="karcher" type="brs_sense" output="screen" />

	<node name="brs_left_wall" pkg="karcher" type="brs_left_wall" output="screen"/>

    <node name="brs_right_wall" pkg="karcher" type="brs_right_wall" output="screen"/>

   <node name="brs_left_step" pkg="karcher" type="brs_left_step" output="screen"/>
    <node name="brs_right_step" pkg="karcher" type="brs_right_step" output="screen"/>

    <node name="brs_pose_data" pkg="karcher" type="brs_pose_data" />
	
    <node name="brs_pose_str" pkg="karcher" type="brs_pose_str" />

    <remap from="/raw_imu" to="imu_data"/>
    <node name="imu_brick_node" pkg="brick_imu_v2" type="imu_brick_node" />

    <node name="brs_teleop" pkg="karcher" type="brs_teleop" output="screen"/>
    <node pkg="rosserial_python" type="serial_node.py" name="serial_node">
    <param name="port" value="/dev/ttyACM0"/>
    <param name="baud" value="57600"/>
  	</node>

    <node name="inc_rplidarNode"          pkg="inclined_lidar"  type="inc_rplidarNode">
	<param name="serial_port"         type="string" value="/dev/inc_rplidar"/>
	<param name="serial_baudrate"     type="int"    value="115200"/>
	<param name="frame_id"            type="string" value="laser"/>
	<param name="inverted"            type="bool"   value="false"/>
	<param name="angle_compensate"    type="bool"   value="true"/>
    </node>
    <node name="rplidar_node" pkg="rplidar_ros" type="rplidarNode">
        <param name="serial_port" type="string" value="/dev/rplidar"/>
        <param name="serial_baudrate" type="int" value="115200"/>
        <param name="frame_id" type="string" value="laser"/>
        <param name="inverted" type="bool" value="false"/>
        <param name="angle_compensate" type="bool" value="true"/>
    </node>
	<!-- param name="PWM_L" value="0.12" type="double" / -->
	<!-- param name="PWM_R" value="0.53" type="double" / -->

	<param name="SET_DISTANCE" value="30.0" type="double" />
	<param name="LEFT_SET" value="0.25" type="double" />
	<param name="RIGHT_SET" value="0.25" type="double" />
	<!-- param name="TURN_TYPE" value="1" type="int" / -->
	
	<node name="Position_control" pkg="karcher" type="Position_control" />
	<node name="Heading_control" pkg="karcher" type="Heading_control" />
	<!-- node name="Turn_control" pkg="karcher" type="Turn_control" / -->
	<node name="Distance_control" pkg="karcher" type="Distance_control" />
	<node name="Navigation_ZZ" pkg="karcher" type="Navigation_ZZ" />

    <param name="/lidar_height"     value="0.80"/><!--Slant lidar, it is set now acc. to mech. drawing-->
    <param name="/safe"             value="2.5"/> <!--do not change-->
    <param name="/turn_dist"        value="950"/> <!--do not change-->
    <param name="/presence_step"    value="4"/><!--do not change-->
    <param name="/presence_wall"    value="1"/><!--do not change-->
    <param name="/sw_param"         value="1.0"/>

    <param name="/obstacle_avoidance"     value="0"/><!-- default is enabled (Only step following)-->
    <param name="/allow_shortwall "     value="0"/><!-- default is disabled (only wall following)-->

    <param name="/ws_setpoint"      value="0.65"/>
    <param name="/ss_enable"        value="0"/> <!--0 is enabled, 1 will be disabled (sensing stopping)-->

    <!--for tuning -->
    <param name="/tuning"           value="1"/> 
    <param name="/KP"               value="0.3"/> <!--decrease if wobbling is more (do not increase)-->
    <param name="/KI"               value="0.0"/> <!--do not change-->
    <param name="/KD"               value="0.15"/> <!--increase to improve alignment-->

    <param name="/new_stop_flat"      value="1"/>
    <param name="/new_stop_slant"     value="1"/>
    <param name="/stop_dist_flat"     value="0.70"/> <!--/scan topic lidar sensing distance-->
    <param name="/stop_dist_slant"    value="0.70"/> <!--/new_scan_far topic lidar sensing distance-->

    <param name="/left_enc_ppr"     value="500"/> <!--left encoder ppr (400 or 500)-->
    <param name="/right_enc_ppr"    value="500"/> <!--right encoder ppr (400 or 500)-->

    <param name="/p_to_p"    value="0"/> <!-- default Pole to pole set at 0 feet -->

    <param name="KP_X"          value="5"/>
    <param name="KD_X"          value="2"/>
    <param name="KP_D"          value="5"/>
    <param name="KD_D"          value="2"/>

    <param name="/lidar_left_end"    value="0.35"/>
    <param name="/lidar_right_end"    value="0.35"/>
</launch>